synapse/storage/databases/main/cache.py

```diff
@@ -223,7 +223,7 @@ def _invalidate_caches_for_event(
         # process triggering the invalidation is responsible for clearing any external
         # cached objects.
         self._invalidate_local_get_event_cache(event_id)
-        self.have_seen_event.invalidate((room_id, event_id))
+        self.have_seen_event.invalidate(((room_id, event_id),))
```

Comment:
This key lookup was wrong and we were never invalidating the `have_seen_event` cache even over replication.

---

Yes, the cache key literally a `set` wrapped in `set`. Something weird with the `TreeCache` I assume

ex. `(('!TnCIJPKzdQdUlIyXdQ:test', '$Iu0eqEBN7qcyF1S9B3oNB3I91v2o5YOgRNPwi_78s-k'),)`

```
LruCache cache values before pop {(('!TnCIJPKzdQdUlIyXdQ:test', '$Iu0eqEBN7qcyF1S9B3oNB3I91v2o5YOgRNPwi_78s-k'),): False}
```

We should probably check all other instances of this for the same problem. And ideally fix the cache so it uses the expected `(room_id, event_id)` key tuple instead.

Reply:
Raised https://github.com/matrix-org/synapse/issues/13865 about this.

Reply:
Used the solution discussed in that issue ⏩